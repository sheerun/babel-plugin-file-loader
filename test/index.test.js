const path = require('path')
const rimraf = require('rimraf')
const fs = require('fs')

const transformCode = require('./transformCode')

function getFixtures (name) {
  return path.resolve(__dirname, 'fixtures', name)
}

describe('index', function () {
  beforeEach(function () {
    rimraf.sync(path.join(__dirname, 'public'))
  })

  it('replaces import statements with uri', function () {
    const result = transformCode(getFixtures('import-image.js'), {}).code
    expect(result).toMatchSnapshot()
  })

  it('handles custom import path', function () {
    const result = transformCode(getFixtures('import-image.js'), {
      publicPath: '/static'
    }).code
    expect(result).toMatchSnapshot()
  })

  it('handles multiple extensions', function () {
    const result = transformCode(getFixtures('import-multiple.js'), {
      publicPath: '/static',
      extensions: ['module.less']
    }).code
    expect(result).toMatchSnapshot()
  })

  it('handles custom import path with trailing slash', function () {
    const result = transformCode(getFixtures('import-image.js'), {
      publicPath: '/static/'
    }).code
    expect(result).toMatchSnapshot()
  })

  it('handles custom name', function () {
    const result = transformCode(getFixtures('import-image.js'), {
      name: '[path][name].[ext]'
    }).code
    expect(result).toMatchSnapshot()
  })

  it('handles custom context', function () {
    const result = transformCode(getFixtures('import-image.js'), {
      context: '/test',
      name: '[path][name].[ext]'
    }).code
    expect(result).toMatchSnapshot()
  })

  it('handles custom context v2', function () {
    const result = transformCode(getFixtures('import-image.js'), {
      context: '/test/assets',
      name: '[path][name].[ext]'
    }).code
    expect(result).toMatchSnapshot()
  })

  it('handles name with custom hash length', function () {
    const result = transformCode(getFixtures('import-image.js'), {
      name: '[hash:8].[ext]'
    }).code
    expect(result).toMatchSnapshot()
  })

  it('handles sha1 hash', function () {
    const result = transformCode(getFixtures('import-image.js'), {
      name: '[sha1:hash].[ext]'
    }).code
    expect(result).toMatchSnapshot()
  })

  it('handles sha256 hash', function () {
    const result = transformCode(getFixtures('import-image.js'), {
      name: '[sha256:hash].[ext]'
    }).code
    expect(result).toMatchSnapshot()
  })

  it('handles sha512 hash', function () {
    const result = transformCode(getFixtures('import-image.js'), {
      name: '[sha512:hash].[ext]'
    }).code
    expect(result).toMatchSnapshot()
  })

  it('handles hex digest', function () {
    const result = transformCode(getFixtures('import-image.js'), {
      name: '[hash:hex].[ext]'
    }).code
    expect(result).toMatchSnapshot()
  })

  it('handles base26 digest', function () {
    const result = transformCode(getFixtures('import-image.js'), {
      name: '[hash:base26].[ext]'
    }).code
    expect(result).toMatchSnapshot()
  })

  it('handles base32 digest', function () {
    const result = transformCode(getFixtures('import-image.js'), {
      name: '[hash:base32].[ext]'
    }).code
    expect(result).toMatchSnapshot()
  })

  it('handles base36 digest', function () {
    const result = transformCode(getFixtures('import-image.js'), {
      name: '[hash:base36].[ext]'
    }).code
    expect(result).toMatchSnapshot()
  })

  it('handles base49 digest', function () {
    const result = transformCode(getFixtures('import-image.js'), {
      name: '[hash:base49].[ext]'
    }).code
    expect(result).toMatchSnapshot()
  })

  it('handles base52 digest', function () {
    const result = transformCode(getFixtures('import-image.js'), {
      name: '[hash:base52].[ext]'
    }).code
    expect(result).toMatchSnapshot()
  })

  it('handles base58 digest', function () {
    const result = transformCode(getFixtures('import-image.js'), {
      name: '[hash:base58].[ext]'
    }).code
    expect(result).toMatchSnapshot()
  })

  it('handles base62 digest', function () {
    const result = transformCode(getFixtures('import-image.js'), {
      name: '[hash:base62].[ext]'
    }).code
    expect(result).toMatchSnapshot()
  })

  it('handles base64 digest', function () {
    const result = transformCode(getFixtures('import-image.js'), {
      name: '[hash:base64].[ext]'
    }).code
    expect(result).toMatchSnapshot()
  })

  it('handles hash as query param', function () {
    const result = transformCode(getFixtures('import-image.js'), {
      name: '[path][name].[ext]?[hash:8]'
    }).code
    expect(result).toMatchSnapshot()
  })

  it('handles full url as publicPath as query param', function () {
    const result = transformCode(getFixtures('import-image.js'), {
      publicPath: 'https://cdn.example.com/project'
    }).code
    expect(result).toMatchSnapshot()
  })

  it('should throw an error when resource does not exist', function () {
    expect(() => {
      transformCode(getFixtures('import-not-existing.js'))
    }).toThrow(/File does not exist/)
  })

  it('should replace require statements with uri', function () {
    const result = transformCode(getFixtures('require-image.js'), {}).code
    expect(result).toMatchSnapshot()
  })

  it('should replace require statements even without assignment', function () {
    const result = transformCode(getFixtures('require-no-var.js'), {}).code
    expect(result).toMatchSnapshot()
  })

  it('should replace require statements in react components', function () {
    const result = transformCode(getFixtures('require-react.js'), {}).code
    expect(result).toMatchSnapshot()
  })

  it('should do nothing when imports have no extensions', function () {
    const result = transformCode(getFixtures('import-no-ext.js')).code
    expect(result).toMatchSnapshot()
  })

  it('should do nothing when require have no extensions', function () {
    const result = transformCode(getFixtures('require-no-ext.js')).code
    expect(result).toEqual(`const test = require('something');`)
  })

  it('should do nothing when not a require assignment', function () {
    const result = transformCode(getFixtures('require-var.js')).code
    expect(result).toEqual(`const test = 'something';`)
  })

  it('outputs file in outputPath', function () {
    transformCode(getFixtures('import-image.js'))
    expect(
      fs.existsSync(
        path.resolve(__dirname, './public/9c87cbf3ba33126ffd25ae7f2f6bbafb.png')
      )
    ).toEqual(true)
  })

  it('doesnt output file when outputPath is null', function () {
    transformCode(getFixtures('import-image.js'), { outputPath: null })
    expect(fs.existsSync(path.resolve(__dirname, './public'))).toEqual(false)
  })

  it('outputs file in outputPath, nested', function () {
    transformCode(getFixtures('import-image.js'), {
      name: '/foo/bar/[hash].[ext]'
    })
    expect(
      fs.existsSync(
        path.resolve(
          __dirname,
          './public/foo/bar/9c87cbf3ba33126ffd25ae7f2f6bbafb.png'
        )
      )
    ).toEqual(true)
  })

  it('outputs file in outputPath, full path', function () {
    transformCode(getFixtures('import-image.js'), {
      name: '[path][name].[ext]'
    })
    expect(
      fs.existsSync(path.resolve(__dirname, './public/test/assets/file.png'))
    ).toEqual(true)
  })

  it('outputs file in outputPath, ignore query', function () {
    transformCode(getFixtures('import-image.js'), {
      name: '[path][name].[ext]?asdfas'
    })
    expect(
      fs.existsSync(path.resolve(__dirname, './public/test/assets/file.png'))
    ).toEqual(true)
  })

  it('inlines file when the length in lower then the limit', function () {
    const result = transformCode(getFixtures('import-text.js'), {
      extensions: ['txt'],
      name: '[path][name].[ext]',
      limit: 6
    }).code
    expect(result).toMatchSnapshot()
  })

  it('doesnt output the file when its inlined', function () {
    transformCode(getFixtures('import-text.js'), {
      extensions: ['txt'],
      name: '[path][name].[ext]',
      limit: 6
    })
    expect(fs.existsSync(path.resolve(__dirname, './public'))).toEqual(false)
  })

  it('doesnt inline file when the length equals the limit', function () {
    const result = transformCode(getFixtures('import-text.js'), {
      extensions: ['txt'],
      name: '[path][name].[ext]',
      limit: 5
    }).code
    expect(result).toMatchSnapshot()
  })

  it('ouputs the file when the length equals the limit', function () {
    transformCode(getFixtures('import-text.js'), {
      extensions: ['txt'],
      name: '[path][name].[ext]',
      limit: 5
    })
    expect(
      fs.existsSync(path.resolve(__dirname, './public/test/assets/file.txt'))
    ).toEqual(true)
  })

  it('inline file with an unknown type by not specifying mime type', function () {
    const result = transformCode(getFixtures('import-unknown.js'), {
      extensions: ['unknown'],
      name: '[path][name].[ext]',
      limit: 6
    }).code
    expect(result).toMatchSnapshot()
  })
})
