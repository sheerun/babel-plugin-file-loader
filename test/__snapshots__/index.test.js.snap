// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`index doesnt inline file when the length equals the limit 1`] = `"const test = \\"/public/test/assets/file.txt\\";"`;

exports[`index handles base26 digest 1`] = `"const test = \\"/public/ccjesbcwzsppzgsqtbnioeblcexw.png\\";"`;

exports[`index handles base32 digest 1`] = `"const test = \\"/public/8vrapkyzxf5ryqy5jmrctwr2ww.png\\";"`;

exports[`index handles base36 digest 1`] = `"const test = \\"/public/ewi6kgnxju4v6e6xs91ffh3mk.png\\";"`;

exports[`index handles base49 digest 1`] = `"const test = \\"/public/wUByBoCTfZTTYAcdFceBFjM.png\\";"`;

exports[`index handles base52 digest 1`] = `"const test = \\"/public/fVWKlApuFbbEMdlLegCqBlW.png\\";"`;

exports[`index handles base58 digest 1`] = `"const test = \\"/public/x5UajugYgSHCpqHbQXrw1d.png\\";"`;

exports[`index handles base62 digest 1`] = `"const test = \\"/public/7F0gARBRl8UjBzQkyqrDyY.png\\";"`;

exports[`index handles base64 digest 1`] = `"const test = \\"/public/3XKCILvWUB_mYicXHPOUus.png\\";"`;

exports[`index handles custom context 1`] = `"const test = \\"/public/assets/file.png\\";"`;

exports[`index handles custom context v2 1`] = `"const test = \\"/public/file.png\\";"`;

exports[`index handles custom import path 1`] = `"const test = \\"/static/9c87cbf3ba33126ffd25ae7f2f6bbafb.png\\";"`;

exports[`index handles custom import path with trailing slash 1`] = `"const test = \\"/static/9c87cbf3ba33126ffd25ae7f2f6bbafb.png\\";"`;

exports[`index handles custom name 1`] = `"const test = \\"/public/test/assets/file.png\\";"`;

exports[`index handles full url as publicPath as query param 1`] = `"const test = \\"https://cdn.example.com/project/9c87cbf3ba33126ffd25ae7f2f6bbafb.png\\";"`;

exports[`index handles hash as query param 1`] = `"const test = \\"/public/test/assets/file.png?9c87cbf3\\";"`;

exports[`index handles hex digest 1`] = `"const test = \\"/public/9c87cbf3ba33126ffd25ae7f2f6bbafb.png\\";"`;

exports[`index handles multiple extensions 1`] = `"const test = \\"/static/934823cbc67ccf0d67aa2a2eeb798f12.module.less\\";"`;

exports[`index handles name with custom hash length 1`] = `"const test = \\"/public/9c87cbf3.png\\";"`;

exports[`index handles sha1 hash 1`] = `"const test = \\"/public/f824ad7c5f655c81c88d3fe267fe5780055bced5.png\\";"`;

exports[`index handles sha256 hash 1`] = `"const test = \\"/public/77c741ac0a9070da1f3c85468cfd7a11532987ad1e9a04631d724629d33c31bb.png\\";"`;

exports[`index handles sha512 hash 1`] = `"const test = \\"/public/a72ac1286999e22d029054119627d677d6ad971132d81926c32284eb379c7f16094d1a0c2f7a281521ecbf6dbd03201b00c5120f985f531d0067b917b4f50709.png\\";"`;

exports[`index inline file with an unknown type by not specifying mime type 1`] = `"const test = \\"data:;base64,MDAwMAo=\\";"`;

exports[`index inlines file when the length in lower then the limit 1`] = `"const test = \\"data:text/plain;base64,MDAwMAo=\\";"`;

exports[`index replaces import statements with uri 1`] = `"const test = \\"/public/9c87cbf3ba33126ffd25ae7f2f6bbafb.png\\";"`;

exports[`index should do nothing when imports have no extensions 1`] = `"import test from 'something';"`;

exports[`index should replace require statements even without assignment 1`] = `"\\"/public/64eababb117f90535085779cc0325003.svg\\";"`;

exports[`index should replace require statements in react components 1`] = `
"const React = require('react');

export default class extends React.Component {
  render() {
    return React.createElement(\\"img\\", {
      src: \\"/public/64eababb117f90535085779cc0325003.svg\\"
    });
  }

}"
`;

exports[`index should replace require statements with uri 1`] = `"const test = \\"/public/64eababb117f90535085779cc0325003.svg\\";"`;
